File: phi_p3dx_mapping/exploration_node.py
import math
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data, QoSProfile, DurabilityPolicy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import OccupancyGrid


class ExplorationNode(Node):
    """
    Classe base para nós de exploração em ROS2.
    Assina mensagens de laser e public a /cmd_vel.
    """
    def __init__(self, node_name: str = 'exploration_node', timer_period: float = 0.1):
        super().__init__(node_name)

        self.laser_ranges = []
        self.laser_angle_min = 0.0
        self.laser_angle_increment = 0.0

        self.map_msg = None

        self.pub_cmd = self.create_publisher(Twist, '/cmd_vel', 10)

        self.create_subscription(
            LaserScan, '/laser_scan', self._cb_laser,
            qos_profile=qos_profile_sensor_data)

        map_qos = QoSProfile(
            depth=1,
            durability=DurabilityPolicy.TRANSIENT_LOCAL
        )
        self.create_subscription(OccupancyGrid, '/map', self._cb_map, map_qos)

        self.create_timer(timer_period, self._control_loop)
        self.get_logger().info(f'[{node_name}] iniciado.')

    def _cb_laser(self, msg: LaserScan) -> None:
        self.laser_ranges = [
            r if msg.range_min <= r <= msg.range_max else float('inf')
            for r in msg.ranges
        ]
        self.laser_angle_min = msg.angle_min
        self.laser_angle_increment = msg.angle_increment
        self.on_laser()

    def _cb_map(self, msg: OccupancyGrid) -> None:
        self.map_msg = msg
        self.on_map()

    def get_front_distance(self, half_angle_deg: float = 15.0) -> float:
        if not self.laser_ranges:
            return float('inf')
        
        n = len(self.laser_ranges)
        center = n // 2
        half = int(math.radians(half_angle_deg) / max(self.laser_angle_increment, 1e-9))
        
        start = max(0, center - half)
        end = min(n - 1, center + half)
        
        region = self.laser_ranges[start : end + 1]
        return min(region) if region else float('inf')

    def publish_velocity(self, v: float, w: float) -> None:
        twist = Twist()
        twist.linear.x = float(v)
        twist.angular.z = float(w)
        self.pub_cmd.publish(twist)

    def stop(self) -> None:
        self.publish_velocity(0.0, 0.0)

    def on_laser(self) -> None:
        pass

    def on_map(self) -> None:
        pass

    def _control_loop(self) -> None:
        pass

File: src/exploration_node.cpp
#include "phi_p3dx_mapping/exploration_node.hpp"
#include <algorithm>
#include <limits>
#include <cmath>

ExplorationNode::ExplorationNode(const std::string &node_name, double timer_period)
  : Node(node_name),
    laser_angle_min_(0.0), laser_angle_increment_(0.0),
    map_msg_(nullptr)
{
  // Publisher: comandos de velocidade
  cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);

  // Subscriber: scan laser
  auto sensor_qos = rclcpp::QoS(rclcpp::SensorDataQoS());
  laser_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
    "laser_scan", sensor_qos,
    std::bind(&ExplorationNode::laser_callback, this, std::placeholders::_1));

  // Subscriber: mapa global
  rclcpp::QoS map_qos(rclcpp::KeepLast(1));
  map_qos.transient_local();
  map_sub_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
    "map", map_qos,
    std::bind(&ExplorationNode::map_callback, this, std::placeholders::_1));

  // Timer para loop de controle
  timer_ = this->create_wall_timer(
    std::chrono::duration<double>(timer_period),
    std::bind(&ExplorationNode::control_loop, this));

  RCLCPP_INFO(this->get_logger(), "[%s] iniciado", node_name.c_str());
}

void ExplorationNode::laser_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
{
  laser_ranges_ = msg->ranges;
  on_laser();
}

void ExplorationNode::map_callback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
{
  map_msg_ = msg;
  on_map();
}

double ExplorationNode::get_front_distance(double half_angle_deg) const
{
  if (laser_ranges_.empty()) return std::numeric_limits<double>::infinity();

  size_t n = laser_ranges_.size();
  size_t center = n / 2;
  size_t half = static_cast<size_t>(std::abs(half_angle_deg * M_PI / 180.0) / std::max(laser_angle_increment_, 1e-9));
  
  size_t start = (center > half) ? center - half : 0;
  size_t end = std::min(center + half, n - 1);

  double min_dist = std::numeric_limits<double>::infinity();
  for (size_t i = start; i <= end; ++i) {
    if (laser_ranges_[i] < min_dist) {
      min_dist = laser_ranges_[i];
    }
  }
  return min_dist;
}

void ExplorationNode::publish_velocity(double v, double w)
{
  geometry_msgs::msg::Twist twist;
  twist.linear.x = v;
  twist.angular.z = w;
  cmd_vel_pub_->publish(twist);
}

void ExplorationNode::stop()
{
  publish_velocity(0.0, 0.0);
}
